sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/cardinality/ApproximateDistinct.java

```diff
@@ -0,0 +1,352 @@
+   * can be estimated thanks to the longest run of starting 0s of the hashed elements.
+   * The longer the run is, the more unlikely it was to happen. Thus, the greater the number
+   * of elements that have been hashed before getting such a run.
+   * <br>Because this algorithm relies mainly on randomness, the stream is divided into buckets
+   * in order to reduce the variance of the estimation.
+   * Therefore, an estimation is applied on several samples and the overall estimation
+   * is then computed using an average (Harmonic mean).
+   *
+   * @param <InputT>    the type of the elements being combined
+   */
+  public static class ApproximateDistinctFn<InputT>
+      extends Combine.CombineFn<InputT, HyperLogLogPlus, HyperLogLogPlus> {
+
+    private final int p;
+
+    private final int sp;
+
+    private ApproximateDistinctFn(int p, int sp) {
+      this.p = p;
+      this.sp = sp;
+    }
+
+      /**
+       * Returns an {@code ApproximateDistinctFn} combiner with the given precision value p.
+       * This means that the input elements will be dispatched into 2^p buckets
+       * in order to estimate the cardinality.
+       *
+       * @param p           precision value for the normal representation
+       * @param <InputT>    the type of the Input {@code Pcollection}'s elements being combined.
+       */
+    public static <InputT> ApproximateDistinctFn<InputT> create(int p) {
+      return new ApproximateDistinctFn<>(p, 0);
+    }
+
+      /**
+       * Returns an {@code ApproximateDistinctFn} combiner with the precision value p
+       * of this combiner and the given precision value sp for the sparse representation,
+       * meaning that the combiner will be using a sparse representation at small cardinalities.
+       *
+       * <p>The sparse representation does not initialize every buckets to 0 at the beggining.
+       * Indeed, for small cardinalities a lot of them will remain empty.
+       * Instead it builds a linked-list that grows when new indexes appear in the hashed values.
+       * To reduce collision of indexes and thus improve precision, we can define {@code sp > p}
+       * When the sparse representation would require more memory than the normal one,
+       * it is converted and the normal algorithm applies for the remaining elements.
+       *
+       * <p><b>WARNING : </b>Choose sp such that {@code p <= sp <= 32}
+       *
+       * <p>Example of use :
+       * <pre>{@code PCollection<Integer> input = ...;
+       * PCollection<HyperLogLogPlus> hllSketch = input
+       *        .apply(Combine.globally(ApproximateDistinct.ApproximateDistinctFn.<Integer>create(p)
+       *                .withSparseRepresentation(sp)));
+       * } </pre>
+       *
+       * @param sp          the precision of HyperLogLog+' sparse representation
+       */
+    public ApproximateDistinctFn<InputT> withSparseRepresentation(int sp) {
+      return new ApproximateDistinctFn<>(this.p, sp);
```

Comment:
<!--new_thread; commit:6bf51fdf6afe6f389415792f335dd08adf9fa85a; resolved:0-->
Perform validation that p <= sp